java-frontend/src/main/java/org/sonar/java/ast/JavaAstScanner.java

```diff
@@ -54,37 +57,48 @@ public JavaAstScanner(ActionParser<Tree> parser, @Nullable SonarComponents sonar
   }
 
   public void scan(Iterable<File> files) {
-    ProgressReport progressReport = new ProgressReport("Report about progress of Java AST analyzer", TimeUnit.SECONDS.toMillis(10));
-    progressReport.start(Lists.newArrayList(files));
-
-    boolean successfullyCompleted = false;
+    if(!files.iterator().hasNext()) {
+      visitor.endOfAnalysis();
+      return;
+    }
+    ExecutorService executor = createExecutor();
+    CompletionService cs =  new ExecutorCompletionService(executor);
     try {
+      int fileSize = 0;
       for (File file : files) {
-        simpleScan(file);
-        progressReport.nextFile();
+        cs.submit(()-> simpleScan(file), null);
+        fileSize++;
       }
-      successfullyCompleted = true;
-    } finally {
-      if (successfullyCompleted) {
-        progressReport.stop();
-      } else {
-        progressReport.cancel();
+      executor.shutdown();
+      for (int i = 0; i < fileSize; i++) {
+        cs.take().get();
+      }
+    } catch (InterruptedException e) {
+      // TODO
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)